### PR TITLE
Asset path fix

### PR DESCRIPTION
**Pull request by tbrebant**
- All files created by the plugin on Android are now placed in a proper subfolder inside the app cache directory, folder called `wizAssets`.
- Database is not removed anymore when requesting the deletion of all assets in cache via `/`.
- Fixed paths problems when passing a file or directory with an uri starting with a slash.

Ping @jrouault for review.

### CODE REVIEW

File: platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java [hunk 1]
```diff
@@ -246,6 +265,12 @@ private void downloadUrl(String fileUrl, String uri, String filePath, CallbackCo
         }
     }
 
+    private boolean createFolderIfRequired(String folderPath) {
+        File folder = new File(folderPath);
+        if (folder.exists()) { return true; }
```
Comment (ronkorving) [inline platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java]: What do you do if exists, but `!folder.isDirectory()`?

File: platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java [hunk 2]
```diff
@@ -67,24 +70,37 @@
...
         super.initialize(cordova, webView);
 
         final Context applicationContext = cordova.getActivity().getApplicationContext();
-        pathToStorage = applicationContext.getCacheDir().getAbsolutePath() + File.separator;
+
+        pathToDatabase = applicationContext.getCacheDir().getAbsolutePath() + File.separator + PLUGIN_FOLDER;
+        pathToAssets = pathToDatabase + File.separator + ASSETS_FOLDER;
+
+        if (!createFolderIfRequired(pathToDatabase)) {
+            Log.e(TAG, "error -- unable to create folder: " + pathToDatabase);
```
Comment (jrouault) [inline platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java]: What happens if the database path creation fails?

File: platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java (hunk 1 above)
Comment (tbrebant) [inline platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java]: You're right, will modify to return `false` in this case.

File: platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetManager.java [hunk 3]
```diff
@@ -197,4 +198,34 @@ public void deleteFolder(String filePath) {
             Log.e(TAG, "Delete file error -- " + e.getMessage(), e);
         }
     }
+
+    // If we detect a data structure created by plugin version <= 5.0.0 we want to clean up
+    private void checkForMigration() {
+        String deprecatedDbLocation = that.getCacheDir().getAbsolutePath() + File.separator + "assets.db";
```
Comment (jrouault) [inline platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetManager.java]: Could it be possible to put `assets.db` and `assets` in two constant at the top of this function? Something such as `DATABASE_NAME_AT_VERSION_500` and `TABLE_NAME_AT_VERSION_500`.

File: platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetManager.java [hunk 4]
```diff
@@ -197,4 +198,34 @@ public void deleteFolder(String filePath) {
...
+        Log.d(TAG, "Found a deprecated database: removing cached files and database");
+        SQLiteDatabase deprecatedDb = SQLiteDatabase.openDatabase(deprecatedDbLocation, null, SQLiteDatabase.OPEN_READONLY);
+        int counter = 0;
+        try {
+            Cursor cursor = deprecatedDb.rawQuery("select * from assets", null);
+            String filePath;
+            while (cursor.moveToNext()) {
+                filePath = cursor.getString(cursor.getColumnIndex("filePath"));
+                File fileToDelete = new File(filePath);
+                if (fileToDelete.exists() && fileToDelete.delete()) {
```
Comment (jrouault) [inline platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetManager.java]: Same as `createFolderIfRequired`, could it be possible to invert the two? Such as:
`fileToDelete.delete() || !fileToDelete.exists()`

File: platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java (hunk 2 above)
Comment (tbrebant) [inline platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java]: Just pushed a new version including an `initialize` method with a callback informing javascript if everything is fine or if the plugin did not initialize and should not be used.

File: platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetManager.java (hunk 3 above)
Comment (tbrebant) [inline platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetManager.java]: ok, will do

File: platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java [hunk 5]
```diff
@@ -101,6 +123,16 @@ private void setBlockSize() {
 
     @Override
     public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
+        if (action.equals(INITIALIZE) && initialized) {
```
Comment (jrouault) [inline platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java]: I would have rather seen something like:
```
if (!initialized) {
    if (action.equals(INITIALIZE_ACTION)) {
        initialize(callbackContext);
    } else {
        callbackContext.error("Plugin failed to initialize");
    }
    return true;
}
```

File: platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java [hunk 6]
```diff
@@ -67,24 +72,41 @@
     private static final int FILE_CREATION_ERROR = 7;
     private static final int JSON_CREATION_ERROR = 8;
 
-    private String pathToStorage;
+    private String pathToDatabase;
+    private String pathToAssets;
     private int blockSize;
 
     @Override
     public void initialize(CordovaInterface cordova, CordovaWebView webView) {
```
Comment (jrouault) [inline platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java]: Why not creating a specific `initialize` method to actually initialize when the `initialize` method is called in JS land?

File: platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java [hunk 7]
```diff
@@ -50,7 +50,12 @@
 
     private String TAG = "WizAssetsPlugin";
     private WizAssetManager wizAssetManager = null;
+    private boolean initialized = false;
 
+    public static final String PLUGIN_FOLDER = "wizAssets";
+    public static final String ASSETS_FOLDER = "assets";
+
+    private static final String INITIALIZE = "initialize";
```
Comment (jrouault) [inline platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java]: Could it be suffixed by `_ACTION` please? 😃

File: platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java [hunk 8]
```diff
@@ -67,24 +72,41 @@
...
+
+        pathToDatabase += File.separator;
+        pathToAssets += File.separator;
+
         setBlockSize();
 
-        wizAssetManager = new WizAssetManager(applicationContext);
+        wizAssetManager = new WizAssetManager(applicationContext, pathToDatabase);
+
+        initialized = wizAssetManager.isReady();
```
Comment (jrouault) [inline platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java]: I think `isReady` is not useful here. If the database could not have been opened, an exception would have been thrown normally no?

File: platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetManager.java [hunk 9]
```diff
@@ -41,28 +41,37 @@
...
             buildDB();
         } else {
-            database = SQLiteDatabase.openDatabase(DATABASE_EXTERNAL_FILE_PATH + File.separator + DATABASE_NAME, null, SQLiteDatabase.OPEN_READWRITE);
+            database = SQLiteDatabase.openDatabase(DATABASE_EXTERNAL_FILE_PATH + DATABASE_NAME, null, SQLiteDatabase.OPEN_READWRITE);
             Log.d(TAG, "DB already initiated.");
         }
     }
 
+    public boolean isReady() {
+        return (new File(DATABASE_EXTERNAL_FILE_PATH + DATABASE_NAME)).exists() && database.isOpen();
```
Comment (tbrebant) [inline platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetManager.java]: As discussed, will also test if database is not `null`

File: platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java (hunk 8 above)
Comment (tbrebant) [inline platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java]: As discussed, ii's safer to keep this test.

File: platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java (hunk 6 above)
Comment (tbrebant) [inline platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java]: As discussed, will rename the API method `isReady`

File: platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java (hunk 5 above)
Comment (tbrebant) [inline platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java]: As discussed, will swap block order

**Comment by tbrebant** [comment]
Updated. I think everything we discussed is included 😃 .

**Comment by tbrebant** [comment]
Updated! Ready for approval :)
